mlflow/R/mlflow/R/tracking-client.R

```diff
@@ -1,33 +1,77 @@
+}
+
+new_mlflow_client_impl <- function(get_host_creds, cli_env = list, clazz = c()) {
   structure(
-    list(
-      tracking_uri = tracking_uri,
-      server_url = server_url %||% tracking_uri
+    list(get_host_creds = get_host_creds,
+         cli_env = cli_env
     ),
-    class = "mlflow_client"
+    class = c(clazz, "mlflow_client")
   )
 }
 
-#' Initialize an MLflow Client
-#'
-#' Initializes an MLflow client.
-#'
-#' @param tracking_uri The tracking URI. If not provided, defaults to the service
-#'  set by `mlflow_set_tracking_uri()`.
-#' @keywords internal
-mlflow_client <- function(tracking_uri = NULL) {
-  tracking_uri <- tracking_uri %||% mlflow_get_tracking_uri()
-  server_url <- if (startsWith(tracking_uri, "http")) {
-    tracking_uri
-  } else if (!is.null(mlflow_local_server(tracking_uri)$server_url)) {
+new_mlflow_host_creds <- function( host = NA, username = NA, password = NA, token = NA,
+                                   insecure = "False") {
+  list(host = host, username = username, password = password, token = token, insecure = insecure)
+}
+
+new_mlflow_client.default <- function(scheme, tracking_uri) {
+  server_url <- if (!is.null(mlflow_local_server(tracking_uri)$server_url)) {
     mlflow_local_server(tracking_uri)$server_url
   } else {
     local_server <- mlflow_server(file_store = tracking_uri, port = mlflow_connect_port())
     mlflow_register_local_server(tracking_uri = tracking_uri, local_server = local_server)
     local_server$server_url
   }
+  new_mlflow_client_impl(get_host_creds = function () {
+    new_mlflow_host_creds(host = server_url)
+  })
+}
+
+basic_http_client <- function(tracking_uri) {
+  get_host_creds <- function () {
+    new_mlflow_host_creds(
+      host = tracking_uri,
+      username = Sys.getenv("MLFLOW_USERNAME", NA),
+      password = Sys.getenv("MLFLOW_PASSWORD", NA),
+      token = Sys.getenv("MLFLOW_TOKEN", NA),
+      insecure = Sys.getenv("MLFLOW_INSECURE", NA)
+    )
+  }
+  cli_env <- function() {
+    res <- list(
+      MLFLOW_USERNAME = Sys.getenv("MLFLOW_USERNAME", NA),
+      MLFLOW_PASSWORD = Sys.getenv("MLFLOW_PASSWORD", NA),
+      MLFLOW_TOKEN = Sys.getenv("MLFLOW_TOKEN", NA),
+      MLFLOW_INSECURE = Sys.getenv("MLFLOW_INSECURE", NA)
+    )
+    res[!is.na(res)]
+  }
+  new_mlflow_client_impl(get_host_creds, cli_env)
+}
+
+new_mlflow_client.http <- function(scheme, tracking_uri) {
+  basic_http_client(tracking_uri)
+}
+
+new_mlflow_client.https <- function(scheme, tracking_uri) {
+  basic_http_client(tracking_uri)
+}
 
-  new_mlflow_client(tracking_uri, server_url = server_url)
+#' Initialize an MLflow Client
+#'
+#' @param tracking_uri The tracking URI. If not provided, defaults to the service
+#'  set by `mlflow_set_tracking_uri()`.
+#' @keywords internal
+mlflow_client <- function(tracking_uri = NULL) {
+  tracking_uri <- tracking_uri %||% mlflow_get_tracking_uri()
+  scheme <- strsplit(tracking_uri, "://") [[1]][1]
+  class(scheme) <- scheme
```

Comment:
Introducing classes is overkill here, let's just switch on `scheme`.

Reply:
Yeah I considered that as well, but I decided to go with classes for better flexibility. What I like about using the classes is that people can extend it without making changes mlflow package. I think it is quite reasonable to expect more custom ways how to fetch auth info from the environment and I don't think we would want to support many  schemes explicitly hardcoded in mlflow package.

Reply:
That's reasonable. Could we then define a constructor for the tracking URI and dispatch `new_mlflow_client` on the URI object? Something along the lines of
```
new_mlflow_uri <- function(uri) {
  class <- paste0("mlflow_uri_", strsplit(uri, "://")[[1]][1])
  structure(uri, class = c(class, "mlflow_uri"))
} 
```
and the signature for `new_mlflow_client()` would become `function(tracking_uri)`, so we can implement e.g. `new_mlflow_client.mlflow_uri_https`. We typically want to avoid `class<-` since it's somewhat hacky.

Reply:
Makes sense, I'll do that. thanks for the suggestion.